storage: merges don't always clear txnwait Queue immediately

SHA: https://github.com/cockroachdb/cockroach/commits/7bc9ea5fbe0c0082fdcfd408245a79c62b00edd4

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestStoreRangeMergeInFlightTxns PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1197346&tab=buildLog

```
=== RUN   TestStoreRangeMergeInFlightTxns
W190325 05:37:56.255784 46655 storage/store.go:1487  [s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I190325 05:37:56.273250 46568 gossip/gossip.go:392  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:40767" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
--- FAIL: TestStoreRangeMergeInFlightTxns (45.41s)
=== RUN   TestStoreRangeMergeInFlightTxns/wait-queue
I190325 05:37:56.532432 47175 storage/replica_command.go:246  [s1,r5/1:/{Systemtse-Table/System…}] initiating a split of this range at key "b" [r23] (manual)
I190325 05:37:56.549591 47175 storage/replica_command.go:385  [s1,r5/1:{/Systemtse-b}] initiating a merge of r23:{b-/Table/SystemConfigSpan/Start} [(n1,s1):1, next=2, gen=0] into this range (manual)
I190325 05:37:56.574647 46645 storage/store.go:2469  [s1,r5/1:{/Systemtse-b},txn=30894eb5] removing replica r23/1
I190325 05:38:11.340550 46662 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range "b" - /Table/SystemConfigSpan/Start that contains live data
I190325 05:38:41.607717 171792 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] kv.TxnCoordSender: heartbeat loop
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190325 05:38:41.607836 171792 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190325 05:38:41.607871 171792 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
    --- FAIL: TestStoreRangeMergeInFlightTxns/wait-queue (45.08s)
        client_merge_test.go:994: context deadline exceeded

```

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -999,7 +999,14 @@ func TestStoreRangeMergeInFlightTxns(t *testing.T) {
 		txn2 := client.NewTxn(ctx, store.DB(), 0 /* gatewayNodeID */, client.RootTxn)
 		txn2ErrCh := make(chan error)
 		go func() {
-			txn2ErrCh <- txn2.Put(ctx, rhsKey, "muhahahah")
+			// Get should block on txn1's intent until txn1 commits.
+			kv, err := txn2.Get(ctx, rhsKey)
+			if err != nil {
+				txn2ErrCh <- err
+			} else if string(kv.ValueBytes()) != t.Name() {
+				txn2ErrCh <- errors.Errorf("actual value %q did not match expected value %q", kv.ValueBytes(), t.Name())
+			}
+			txn2ErrCh <- nil
 		}()
 
 		// Wait for txn2 to realize it conflicts with txn1 and enter its wait queue.
@@ -1030,21 +1037,14 @@ func TestStoreRangeMergeInFlightTxns(t *testing.T) {
 			t.Fatal(err)
 		}
 
-		kv, pErr := store.DB().Get(ctx, rhsKey)
-		if pErr != nil {
-			t.Fatal(pErr)
-		} else if string(kv.ValueBytes()) != t.Name() {
-			t.Fatalf("actual value %q did not match expected value %q", kv.ValueBytes(), t.Name())
-		}
-
-		// Now that txn1 has committed, txn2's put operation should complete.
+		// Now that txn1 has committed, txn2's get operation should complete.
 		select {
 		case err := <-txn2ErrCh:
 			if err != nil {
 				t.Fatal(err)
 			}
 		case <-ctx.Done():
-			t.Fatal("timed out waiting for txn2 to complete put")
+			t.Fatal("timed out waiting for txn2 to complete get")
 		}
 
 		if err := txn2.Commit(ctx); err != nil {

